python/cuml/ts/arima.pyx

```diff
@@ -0,0 +1,814 @@
+    cdef vector[double] vec_Tma
+
+    pynvtx_range_push("batched_transform")
+    
+    if handle is None:
+        handle = cuml.common.handle.Handle()
+    cdef cumlHandle* handle_ = <cumlHandle*><size_t>handle.getHandle()
+    cdef np.ndarray[double] Tx = np.zeros(nb*(d+p+q))
+
+    batched_jones_transform(handle_[0], p, d, q, nb, isInv, &x[0], &Tx[0])
+
+    pynvtx_range_pop()
+    return (Tx)
+
+
+
+def _start_params(order, y_diff):
+    """A quick approach to determine reasonable starting mu (trend), AR, and MA parameters"""
+
+    # y is mutated so we need a copy
+    y = np.copy(y_diff)
+    nobs = len(y)
+
+    p, q, d = order
+    params_init = np.zeros(p+q+d)
+    if d > 0:
+        # center y (in `statsmodels`, this is result when exog = [1, 1, 1...])
+        mean_y = np.mean(y)
+        params_init[0] = mean_y
+        y -= mean_y
+
+    if p == 0 and q == 0:
+        return params_init
+
+    if p != 0:
+
+        # TODO: `statsmodels` uses BIC to pick the "best" `p` for this initial
+        # fit. The "best" model is probably a p=1, so we will assume that for now.
+        p_best = 1
+        x = np.zeros((len(y) - p_best, p_best))
+        # create lagged series set
+        for lag in range(1, p_best+1):
+            # create lag and trim appropriately from front so they are all the same size
+            x[:, lag-1] = y[p_best-lag:-lag].T
+
+        # LS fit a*X - Y
+        y_ar = y[p_best:]
+        
+        (ar_fit, _, _, _) = np.linalg.lstsq(x, y_ar.T, rcond=None)
+        # print("initial_ar_fit:", ar_fit)
+        # set_trace()
+        if q == 0:
+            params_init[d:] = ar_fit
+        else:
+            residual = y[p_best:] - np.dot(x, ar_fit)
+
+            assert p >= p_best
+            p_diff = p - p_best
+
+            x_resid = np.zeros((len(residual) - q - p_diff, q))
+            x_ar2 = np.zeros((len(residual) - q - p_diff, p))
+ 
+            # create lagged residual and ar term
+            for lag in range(1, q+1):
+                x_resid[:, lag-1] = (residual[q-lag:-lag].T)[p_diff:]
+            for lag in range(1, p+1):
+                x_ar2[:, lag-1] = (y[p-lag:-lag].T)[q:]
+
+            X = np.column_stack((x_ar2, x_resid))
+            (arma_fit, _, _, _) = np.linalg.lstsq(X, y_ar[(q+p_diff):].T, rcond=None)
+
+            params_init[d:] = arma_fit
+
+    else:
+        # case when p == 0 and q>0
+
+        # when p==0, MA params are often -1
+        # TODO: See how `statsmodels` handles this case
+        params_init[d:] = -1*np.ones(q)
+
+    return params_init
+
+
+
+def _model_complexity(order):
+    (p, d, q) = order
+    # complexity is number of parameters: mu + ar + ma
+    return d + p + q
+
+
+def _batch_trans(p, d, q, nb, x, handle=None):
+    """Apply the stationarity/invertibility guaranteeing transform to batched-parameter vector x."""
+    pynvtx_range_push("jones trans")
+
+    if handle is None:
+        handle = cuml.common.handle.Handle()
```

Comment:
hey here is what I was suggesting for `fit` :). I guess it was just an oversight since its in all of this methods 👍